### PR TITLE
BOLT 7: announcement_signature. Type: 259

### PR DESCRIPTION
**Pull request by kushagra0902**
- Adds BOLT 7 announcement_signatures support (type 259, 0x0103) with full encode and decode coverage.
- Wires the message into the top-level BOLT dispatcher and exports.
- Adds and uses typed primitives plus wire-format support for ShortChannelId and gossip Signature.
 
To test: 

- cargo test -p smite announcement_signature::tests -- --nocapture
- cargo test -p smite message_announcement_signatures_roundtrip -- --nocapture

### CODE REVIEW

File: smite/src/bolt/types.rs [hunk 1]
```diff
@@ -39,6 +42,42 @@ impl ChannelId {
...
+    /// Returns the encoded 64-bit representation.
+    #[must_use]
+    pub const fn value(self) -> u64 {
+        self.0
+    }
+}
+
+/// A 64-byte compact gossip signature.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Signature(pub [u8; SIGNATURE_SIZE]);
```
Comment (Vineet1101) [inline smite/src/bolt/types.rs]: Hey I found an issue in your code this custom wrapper completely bypasses standard cryptographic validation and introduces a severe regression. 

as per the [BOLT 7 `announcement_signatures` spec](https://github.com/lightning/bolts/blob/master/07-routing-gossip.md#the-announcement_signatures-message), these fields are standard compact ECDSA signatures. We already have `secp256k1::ecdsa::Signature` integrated into this codebase, which uses `.from_compact()` to mathematically validate that the r and s values are valid points on the curve during decoding. 

By defining this as a raw byte array, you are instructing the node to blindly accept mathematically invalid/garbage signatures off the wire. 
Sometimes llm generated code does this as it does not have  full context of the spec and just to make compiler happy it does this :)

File: smite/src/bolt/announcement_signature.rs [hunk 2]
```diff
@@ -0,0 +1,166 @@
+//! BOLT 7 `announcement_signatures` message.
+
+use super::BoltError;
+use super::types::{ChannelId, ShortChannelId, Signature};
+use super::wire::WireFormat;
+
+/// BOLT 7 `announcement_signatures` message (type 259 / 0x0103).
```
Comment (ekzyis) [inline smite/src/bolt/announcement_signature.rs:7]:
```suggestion
/// BOLT 7 `announcement_signatures` message (type 259).
```

File: smite/src/bolt/types.rs [hunk 3]
```diff
@@ -17,6 +17,9 @@ pub const TXID_SIZE: usize = 32;
 /// Size of a compact ECDSA signature in bytes.
 pub const COMPACT_SIGNATURE_SIZE: usize = 64;
 
+/// Size of a compact gossip signature in bytes.
+pub const SIGNATURE_SIZE: usize = COMPACT_SIGNATURE_SIZE;
```
Comment (ekzyis) [inline smite/src/bolt/types.rs:21]: I don't think we need to distinguish signatures by the context in which they are used in. It adds a layer of confusion: "What is a compact _gossip_ signature? Is it different from compact ECDSA signatures?"

File: smite/src/bolt/announcement_signature.rs (hunk 2 above)
Comment (morehouse) [inline smite/src/bolt/announcement_signature.rs:7]: Omit the hex, no other codec does that.
```suggestion
/// BOLT 7 `announcement_signatures` message (type 259).
```

File: smite/src/bolt/announcement_signature.rs [hunk 4]
```diff
@@ -0,0 +1,166 @@
+//! BOLT 7 `announcement_signatures` message.
```
Comment (morehouse) [inline smite/src/bolt/announcement_signature.rs:1]: The file name should be `announcement_signatures.rs`.

File: smite/src/bolt/types.rs [hunk 5]
```diff
@@ -39,6 +42,42 @@ impl ChannelId {
...
+    pub const fn value(self) -> u64 {
+        self.0
+    }
+}
+
+/// A 64-byte compact gossip signature.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Signature(pub [u8; SIGNATURE_SIZE]);
+
+impl Signature {
```
Comment (ekzyis) [inline smite/src/bolt/types.rs:67]: see https://github.com/morehouse/smite/pull/25#discussion_r3040731572

File: smite/src/bolt/types.rs [hunk 6]
```diff
@@ -39,6 +42,42 @@ impl ChannelId {
     }
 }
 
+/// Compact channel identifier containing block, tx, and output indexes.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct ShortChannelId(pub u64);
+
+impl ShortChannelId {
```
Comment (ekzyis) [inline smite/src/bolt/types.rs:49]: I can see how it could make sense to distinguish short channel IDs from u64 in the future, but for now, I agree with https://github.com/morehouse/smite/pull/25#discussion_r3040742587:

> Let's just use `u64` for simplicity?

File: smite/src/bolt/types.rs (hunk 3 above)
Comment (morehouse) [inline smite/src/bolt/types.rs:21]: Seems pointless -- just use `COMPACT_SIGNATURE_SIZE`.

File: smite/src/bolt/announcement_signature.rs
Comment (ekzyis) [inline smite/src/bolt/announcement_signature.rs:1]: I would rename this to announcement_signatures.rs to match message name

File: smite/src/bolt/types.rs [hunk 7]
```diff
@@ -39,6 +42,42 @@ impl ChannelId {
     }
 }
 
+/// Compact channel identifier containing block, tx, and output indexes.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct ShortChannelId(pub u64);
```
Comment (morehouse) [inline smite/src/bolt/types.rs:47]: This newtype doesn't seem useful.  It just wraps `u64` with no new functionality.

It would be more useful to add methods to build from or extract the 3 fields (block, tx, output).

File: smite/src/bolt/types.rs (hunk 5 above)
Comment (morehouse) [inline smite/src/bolt/types.rs:67]: Just use `secp256k1::ecdsa::Signature` like every other codec.

**Review by morehouse — CHANGES_REQUESTED** [review]
Please squash into one commit and make sure CI passes before requesting another review.

File: smite/src/bolt/announcement_signature.rs [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AnnouncementSignatures {
+    /// The channel this message refers to.
+    pub channel_id: ChannelId,
+    /// The compact on-chain identifier of the channel's funding output.
+    pub short_channel_id: ShortChannelId,
+    /// The sender's node signature over the `channel_announcement`.
+    pub node_signature: Signature,
+    /// The sender's bitcoin key signature over the `channel_announcement`.
+    pub bitcoin_signature: Signature,
```
Comment (ekzyis) [inline smite/src/bolt/announcement_signature.rs:24]: see https://github.com/morehouse/smite/pull/25#discussion_r3040731572

File: smite/src/bolt/announcement_signature.rs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    ///
+    /// # Errors
+    ///
+    /// Returns `Truncated` if the payload is too short for any field.
+    pub fn decode(payload: &[u8]) -> Result<Self, BoltError> {
+        let mut cursor = payload;
+        let channel_id       = ChannelId::read(&mut cursor)?;
+        let short_channel_id = ShortChannelId::read(&mut cursor)?;
+        let node_signature   = Signature::read(&mut cursor)?;
+        let bitcoin_signature = Signature::read(&mut cursor)?;
```
Comment (ekzyis) [inline smite/src/bolt/announcement_signature.rs:49]: could use `WireFormat::read` for consistency with (most) other code

File: smite/src/bolt/announcement_signature.rs [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use super::super::{CHANNEL_ID_SIZE, SIGNATURE_SIZE};
+
+    fn dummy() -> AnnouncementSignatures {
```
Comment (ekzyis) [inline smite/src/bolt/announcement_signature.rs:65]: see other code for how this is usually named

File: smite/src/bolt/announcement_signature.rs [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+        }
+    }
+
+    #[test]
+    fn encode_field_sizes() {
+        let encoded = dummy().encode();
+        // channel_id(32) + short_channel_id(8) + node_sig(64) + bitcoin_sig(64)
+        assert_eq!(
+            encoded.len(),
+            CHANNEL_ID_SIZE + 8 + SIGNATURE_SIZE + SIGNATURE_SIZE
```
Comment (ekzyis) [inline smite/src/bolt/announcement_signature.rs:80]: see other code

```suggestion
            168
```

File: smite/src/bolt/announcement_signature.rs [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+    fn roundtrip_different_values() {
+        let msg = AnnouncementSignatures {
+            channel_id:        ChannelId::new([0xff; CHANNEL_ID_SIZE]),
+            short_channel_id:  ShortChannelId::new(0xffff_ffff_ffff_ffff),
+            node_signature:    Signature::new([0xab; SIGNATURE_SIZE]),
+            bitcoin_signature: Signature::new([0xcd; SIGNATURE_SIZE]),
+        };
+        let decoded = AnnouncementSignatures::decode(&msg.encode()).unwrap();
+        assert_eq!(msg, decoded);
+    }
```
Comment (ekzyis) [inline smite/src/bolt/announcement_signature.rs:101]: I don't see why we need two roundtrip tests here. Other messages only have one roundtrip test.

File: smite/src/bolt/announcement_signature.rs [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    fn short_channel_id_roundtrip_boundary_values() {
+        for scid in [0u64, 1, u64::MAX] {
+            let msg = AnnouncementSignatures {
+                short_channel_id: ShortChannelId::new(scid),
+                ..dummy()
+            };
+            let decoded = AnnouncementSignatures::decode(&msg.encode()).unwrap();
+            assert_eq!(decoded.short_channel_id, msg.short_channel_id);
+        }
+    }
```
Comment (ekzyis) [inline smite/src/bolt/announcement_signature.rs:165]: I think this additional roundtrip test is also not needed, even though I see it tests value boundaries.

**Review by ekzyis — CHANGES_REQUESTED** [review]

**Comment by ekzyis** [comment]
Oh, I'm too slow, I didn't see @morehouse's review before I finished my review haha

**Comment by morehouse** [comment]
Superseded by #121.
